SMLHelper/Patchers/PingTypePatcher.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Harmony;
+using SMLHelper.V2.Handlers;
+using SMLHelper.V2.Utility;
+
+namespace SMLHelper.V2.Patchers
+{
+    internal class PingTypePatcher
+    {
+        private const string TechTypeEnumName = "PingType";
+        internal static readonly int startingIndex = 10;
+        internal static readonly List<int> bannedIndices = new List<int>();
+        private static readonly Dictionary<PingType, Atlas.Sprite> cachedSprites = new Dictionary<PingType, Atlas.Sprite>();
+
+        internal static readonly EnumCacheManager<PingType> cacheManager =
+            new EnumCacheManager<PingType>(
+                enumTypeName: TechTypeEnumName,
+                startingIndex: startingIndex,
+                bannedIDs: ExtBannedIdManager.GetBannedIdsFor(TechTypeEnumName, bannedIndices, PreRegisteredCraftTreeTypes()));
+
+        private static List<int> PreRegisteredCraftTreeTypes()
+        {
+            var declaredPingTypes = (PingType[]) Enum.GetValues(typeof(PingType));
+            var bannedIndices = declaredPingTypes.Select(t => (int) t)
+                .Where(t => t > startingIndex)
+                .Distinct()
+                .ToList();
+            
+            Logger.Log($"Finished known PingType exclusion. {bannedIndices.Count} IDs were added in ban list.");
+            return bannedIndices;
+        }
+
+        internal static PingType AddPingType(string name, Atlas.Sprite sprite)
+        {
+            var cache = cacheManager.RequestCacheForTypeName(name) ?? new EnumTypeCache()
+            {
+                Name = name,
+                Index = cacheManager.GetNextAvailableIndex()
+            };
+
+            if (cacheManager.IsIndexAvailable(cache.Index))
+                cache.Index = cacheManager.GetNextAvailableIndex();
+
+            var pingType = (PingType) cache.Index;
+            cacheManager.Add(pingType, cache.Index, cache.Name);
+            
+            cachedSprites.Add(pingType, sprite);
+            if (PingManager.sCachedPingTypeStrings.valueToString.ContainsKey(pingType) == false)
+                PingManager.sCachedPingTypeStrings.valueToString.Add(pingType, name);
+
+            if (PingManager.sCachedPingTypeTranslationStrings.valueToString.ContainsKey(pingType) == false)
+                PingManager.sCachedPingTypeTranslationStrings.valueToString.Add(pingType, name);
+            
+            Logger.Log($"Successfully added PingType: '{name}' to Index: '{cache.Index}'", LogLevel.Debug);
+            return pingType;
+        }
+
+        internal static PingType? GetPingType(string pingName)
+        {
+            try
+            {
+                var ping = Enum.Parse(typeof(PingType), pingName);
+                return (PingType) ping;
```

Comment:
Can we use use [Enum.TryParse](https://docs.microsoft.com/en-us/dotnet/api/system.enum.tryparse?view=netframework-4.0) instead of a try-catch block?  
Also use the case-insensitive version.

Reply:
We're calling at this point a patched version of the Parse method so ignoreCase wouldnt do anything. Im not using TryParse here as TryParse isn't patched only Parsed. Have you got an example of harmony patching TryParse?

Reply:
I see what you're doing here now. I was confused because there isn't an analogous method to this in either `TechTypePatcher` or `CraftTreeTypePatcher`.

The problem here is that this implementation relies on a Harmony patch to work, but any mod that requests this info will be using it _before_ SMLHelper has made its Harmony patches.

If you want to expose functionality that lets external consumers get the `PingType` enum value for a given ping name string, use something like the `TryGetModdedTechType` and `ModdedTechTypeExists` methods of the [TechTypeHandler](https://github.com/SubnauticaModding/SMLHelper/blob/master/SMLHelper/Handlers/TechTypeHandler.cs), which make use of the enum cache.  
If you really want to be able to handle vanilla values as well, then you can use a switch statement like [ValidateStandardCraftTree](https://github.com/SubnauticaModding/SMLHelper/blob/master/SMLHelper/Handlers/CraftTreeHandler.cs#L367) to go through all the known types before checking the cache on the `default` case.